test/middleware-test.ts

```diff
@@ -0,0 +1,696 @@
+// /test/middleware-test.ts
+// Tests the rate limiting middleware
+
+import { jest } from '@jest/globals'
+import Express from 'express'
+import request from 'supertest'
+
+import rateLimit, {
+	Store,
+	Options,
+	IncrementResponse,
+} from '../source/index.js'
+
+import { createServer } from './helpers/create-server.js'
+
+describe('middleware test', () => {
+	afterEach(() => {
+		jest.restoreAllMocks()
+	})
+
+	class MockStore implements Store {
+		initWasCalled = false
+		incrementWasCalled = false
+		decrementWasCalled = false
+		resetKeyWasCalled = false
+
+		counter = 0
+
+		init(_options: Options): void {
+			this.initWasCalled = true
+		}
+
+		async increment(_key: string): Promise<IncrementResponse> {
+			this.counter += 1
+			this.incrementWasCalled = true
+
+			return { totalHits: this.counter, resetTime: undefined }
+		}
+
+		async decrement(_key: string): Promise<void> {
+			this.counter -= 1
+			this.decrementWasCalled = true
+		}
+
+		async resetKey(_key: string): Promise<void> {
+			this.resetKeyWasCalled = true
+		}
+	}
+
+	it('should not modify the options object passed', () => {
+		const options = {}
+		rateLimit(options)
+		expect(options).toStrictEqual({})
+	})
+
+	it('should call `init` even if no requests have come in', async () => {
+		const store = new MockStore()
+		rateLimit({
+			store,
+		})
+
+		expect(store.initWasCalled).toEqual(true)
+	})
+
+	it('should print an error if `request.ip` is undefined', async () => {
+		jest.spyOn(global.console, 'error').mockImplementation(() => {})
+
+		await Promise.resolve(
+			rateLimit()(
+				{ ip: undefined } as any as Express.Request,
+				{} as any as Express.Response,
+				(() => {}) as Express.NextFunction,
+			),
+		)
+
+		expect(console.error).toBeCalled()
+	})
+
+	it('should let the first request through', async () => {
+		const app = createServer(rateLimit({ max: 1 }))
+
+		await request(app).get('/').expect(200).expect('Hi there!')
+	})
+
+	it('should call `increment` on the store', async () => {
+		const store = new MockStore()
+		const app = createServer(
+			rateLimit({
+				store,
+			}),
+		)
+		await request(app).get('/')
+
+		expect(store.incrementWasCalled).toEqual(true)
+	})
+
+	it('should call `resetKey` on the store', async () => {
+		const store = new MockStore()
+		const limiter = rateLimit({
+			store,
+		})
+
+		limiter.resetKey('key')
+
+		expect(store.resetKeyWasCalled).toEqual(true)
+	})
+
+	describe('timer tests', () => {
+		beforeEach(() => {
+			jest.useFakeTimers('modern')
+		})
+		afterEach(() => {
+			jest.useRealTimers()
+		})
```

Comment:
I think we might as well move these to the top-level `describe` block - the clock was mocked for every test in this file previously, and while it doesn't make any difference for the ones with the MockStore, it prevents leaking `setInterval`s for the ones that use the "real" MemoryStore.

Reply:
(We should probably also do that for the MemoryStore tests...)

Reply:
Done